browser/ui/webui/settings/brave_import_bulk_data_handler.cc

```diff
@@ -0,0 +1,137 @@
+/* Copyright (c) 2023 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/browser/ui/webui/settings/brave_import_bulk_data_handler.h"
+
+#include <memory>
+#include <string>
+#include <unordered_map>
+#include <utility>
+
+#include "base/bind.h"
+#include "base/files/file_path.h"
+#include "base/files/file_util.h"
+#include "base/json/json_reader.h"
+#include "base/rand_util.h"
+#include "base/strings/utf_string_conversions.h"
+#include "chrome/browser/browser_process.h"
+#include "chrome/browser/profiles/profile_avatar_icon_util.h"
+#include "chrome/browser/profiles/profile_manager.h"
+#include "chrome/common/pref_names.h"
+#include "components/prefs/pref_service.h"
+
+namespace settings {
+
+BraveImportBulkDataHandler::BraveImportBulkDataHandler() = default;
+BraveImportBulkDataHandler::~BraveImportBulkDataHandler() = default;
+
+void BraveImportBulkDataHandler::RegisterMessages() {
+  BraveImportDataHandler::RegisterMessages();
+  web_ui()->RegisterMessageCallback(
+      "importDataBulk",
+      base::BindRepeating(&BraveImportBulkDataHandler::HandleImportDataBulk,
+                          base::Unretained(this)));
+}
+
+void BraveImportBulkDataHandler::PrepareProfile(
+    const std::u16string& name,
+    ProfileReadyCallback profile_ready_callback) {
+  ProfileManager* profile_manager = g_browser_process->profile_manager();
+  Profile* profile =
+      profile_manager->GetProfileByPath(profile_manager->user_data_dir().Append(
+          base::FilePath::FromASCII(base::UTF16ToUTF8(name))));
+
+  if (profile) {
+    std::move(profile_ready_callback).Run(profile);
+    return;
+  }
+
+  auto avatar_index = base::RandInt(profiles::GetModernAvatarIconStartIndex(),
+                                    profiles::GetDefaultAvatarIconCount());
+  ProfileManager::CreateMultiProfileAsync(
+      name, avatar_index, false,
+      base::BindOnce(
+          [](ProfileReadyCallback initialized_callback, Profile* profile) {
+            CHECK(profile);
+            // Migrate welcome page flag to new profiles.
+            profile->GetPrefs()->SetBoolean(prefs::kHasSeenWelcomePage, true);
+            std::move(initialized_callback).Run(profile);
+          },
+          std::move(profile_ready_callback)));
+}
+
+void BraveImportBulkDataHandler::ProfileReadyForImport(
```

Comment:
If profile is ready but doesn't have disk access, what happens?
I assume import process is stoped with profile creation.
If so, how about checking disk access first and then doing profile prepare?

Reply:
done